### ISSUE TITLE
Introduce `#[builder(getter)]`

### ISSUE THREAD

**Issue by Veetaha** [elastio/bon#225]
This is a tracking issue for the `#[builder(getter)]` attribute feature. The stabilization of the design for this attribute requires some more feedback from the community.

If you think the current design is already solid and covers your current or potential future use cases, then just leave a 👍 reaction.

If you'd like to change something about the `#[builder(getter)]` attribute syntax and behavior or just extend it, then feel free to leave a comment. We'd be glad to hear your ideas!

### A note for the community from the maintainers

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to help the maintainers with prioritizing it. You may add a comment describing your real use case related to this issue for us to better understand the problem domain.

**Comment by Veetaha** [comment]
@angelorendina you mentioned in https://github.com/elastio/bon/issues/149#issuecomment-2523194682 that you got `getter(mut)` working. Although, I'd expect that to be difficult to do because `mut` is a keyword, and [`darling`](https://docs.rs/darling/latest/darling/) doesn't support parsing keywords due to syn's limitations (https://github.com/dtolnay/syn/issues/1458).

Did you implement custom parsing of `MetaList` in your implementation?

**Comment by angelorendina** [comment]
@Veetaha in my implementation attempt, I simply used `getter_mut` as attribute 😄
I got it working with a simple example (essentially a copy of what I've found in the playground and tests folder) but I haven't given much thought on how the feature would interact with other attributes or functionalities, so I cannot guarantee there are no inconsistencies or pitfalls.

**Comment by Veetaha** [comment]
Yeah, `getter_mut` is what I was expecting to see, since that would not fall into the `darling` parsing trap.

Just a headsup, I have a draft PR, that I'm currently working on for `deref, copy, clone` support (https://github.com/elastio/bon/pull/229). I added a custom `FromMeta` implementation in that PR for `GetterConfig`. I think once that one is merged it should be easy to add custom `getter(mut)` parsing in that `FromMeta` impl.

**Comment by Veetaha** [comment]
I'm revisiting the getters design for `start_fn` members and `self` receiver of methods. I think we can avoid providing getters for them, and make those available as regular private fields on the builder. This aligns their behavior with `#[builder(field)]`. Also, providing them as regular fields is more convenient than providing getters. E.g. the following should be possible:

```rust
#[derive(bon::Builder)]
struct Command {
    #[builder(start_fn)]
    args: Vec<String>,
}

impl<S: command_builder::State> CommandBuilder<S> {
    fn arg(mut self, arg: String) -> Self {
        self.args.push(arg);
        self
    }
}
```

The dilemma here is in the naming of the builder's field for the method's `self` receiver. Rust disallows `r#self` identifier, so how should it be called instead? 
- `self_`
- `receiver`
- `this`

**Comment by musjj** [comment]
I think one thing I would love to see is the ability to generate getters for all fields directly:

```rust
#[derive(Builder)]
#[builder(getter)]
struct User {
    name: String,
    is_admin: bool,
    level: Option<u32>,
}
```

This currently generates an error.

**Comment by Veetaha** [comment]
I think having `on(_, getter)` makes sense, although I wonder what is your use case for getters on builders, @musjj?

derive(Builder) isn't going to generate getters on the built struct, and this is by design. Otherwise it would be too surprising to see some other methods generated on the struct by derive(Builder) other than `builder()`. There is an open issue about deriving accessors (#147) where that could fit. However l haven't even decided wether to put those into the same or separate crate and how to organize the docs for that to make it coexist with the builder feature

**Comment by musjj** [comment]
Yeah, I guess it's out-of-scope-ish for this crate. I just found out there's already a crate for this: https://github.com/jbaublitz/getset and it works well together with bon.

**Comment by TheButlah** [comment]
Posting here since you asked for feedback. I am brand new to bon. Here is my use case:

Builder roughly looks like:

```rust
struct Cfg {
  a: A,
  b: B,
}
```

I want to allow the user to either set `b` directly themselves, or "derive it" from the value of `a`. So something like:

```rust
impl<S: cfg_builder::State> CfgBuilder<S> {
    pub fn b_from_a(
        self,
    ) -> CfgBuilder<cfg_builder::SetB<S>>
    where
        S::A: cfg_builder::IsSet,
        S::B: cfg_builder::IsUnset,
    {
        let a = &self.a; // not possible! Unless maybe with the getters api?
        let b = {
            // some code that computes a `B` from `a`.
        };
        self.b(b)
    }
}
```

Importantly, the getter for `a` should not be publicly visible, its just for internal use by my custom setter.

Is this something that the getter feature would be intended for?

**Comment by MareStare** [comment]
Hi @TheButlah, I think for your use case it may be enough to set a default for field `b` that depends on the value of field `a` like this:

```rust
#[derive(bon::Builder)]
struct Cfg {
  a: u32,

  #[builder(default = a * 2)]
  b: u32,
}
fn main() {
    // Derive the value of `b` from `a` if `b` is not provided
    let cfg = Cfg::builder().a(3).build();
    assert_eq!(cfg.b, 6);

    // Use the provided value of `b` if it is provided
    let cfg = Cfg::builder().a(3).b(5).build();
    assert_eq!(cfg.b, 5);
}
```

For more advanced use cases I recommend defining a builder based on a function ([function builder paradigm shift](https://bon-rs.com/guide/alternatives#function-builder-paradigm-shift)).

Does this solution fit your problem?

**Comment by Veetaha** [comment]
Hi @TheButlah, I think for your use case it may be enough to set a default for field `b` that [depends](https://bon-rs.com/reference/builder/member/default#evaluation-context) on the value of field `a` like this:

```rust
#[derive(bon::Builder)]
struct Cfg {
  a: u32,

  #[builder(default = a * 2)]
  b: u32,
}
fn main() {
    // Derive the value of `b` from `a` if `b` is not provided
    let cfg = Cfg::builder().a(3).build();
    assert_eq!(cfg.b, 6);

    // Use the provided value of `b` if it is provided
    let cfg = Cfg::builder().a(3).b(5).build();
    assert_eq!(cfg.b, 5);
}
```

For more advanced use cases I recommend defining a builder based on a function ([function builder paradigm shift](https://bon-rs.com/guide/alternatives#function-builder-paradigm-shift)).

Does this solution fit your problem?

**Comment by TheButlah** [comment]
I think this would have worked, but in reality, my `b_from_a` is an async function. AFAICT async is not supported in the expression passed to the `builder(default)` attribute.

I just read the page you linked about the function builder paradigm shift, but I didn't understand how this applies here. Can you elaborate?

**Comment by TheButlah** [comment]
More generally though, I think it is reasonable that custom setters should get access to any prior initialized values, via the `IsSet` marker trait. Or at least access to the internal `Option`s.

**Comment by Veetaha** [comment]
@TheButlah I see, the fact that your `b_from_a` should be async changes everything. The function-builder approach can still work here like this, although your builder would then be uncoditionally async:

```rust
struct Cfg {
    a: u32,
    b: u32,
}

#[bon::bon]
impl Cfg {
    #[builder]
    async fn new(a: u32, b: Option<u32>) -> Self {
        let b = if let Some(b) = b {
            b
        } else {
            async { a * 2 }.await
        };

        Self { a, b }
    }
}

#[tokio::main]
async fn main() {
    // Derive the value of `b` from `a` if `b` is not provided
    let cfg = Cfg::builder().a(3).build().await;
    assert_eq!(cfg.b, 6);

    // Use the provided value of `b` if it is provided
    // Unfortunately, the `build()` in this case is still async
    let cfg = Cfg::builder().a(3).b(5).build().await;
    assert_eq!(cfg.b, 5);
}
```

However, if you want to be more granular and have a sync signature you do need the getter feature to implement a custom *async setter* as you shown.

So now I get your point, the `#[builder(getter)]` feature does solve your use case. Note that we can't expose the underlying fields of the builder (they are considered private), because the builder does an unsafe `unwrap_unchecked()` in its finishing function for required fields because it relies on `IsSet` type state to guarantee that internal `Option<_>`s are `Some`.

**Comment by TheButlah** [comment]
OK, I will upvote the original post not because I actually want end-users to have access to the getters, but because I want access to the getters in my custom setters :P

**Comment by Veetaha** [comment]
Understandable, using getters only privately in setters is one of the use cases I envisioned. Just make sure to mark the getters as private with `#[builder(getter(vis = ""))]` or `#[builder(getter(vis = "pub(self)"))]`

**Comment by Veetaha** [comment]
I'm stabilizing the current MVP of this feature that's been in experimental state for 3 months: https://github.com/elastio/bon/pull/251.

The next minor version of `bon` `3.4`, that I'll release today, will no longer require the cargo feature `experimental-getters` to use the existing `#[builder(getter)]` API.

I've renamed this issue to denote that what's left is to implement the rest of the batteries design as described in the issue body: 
- mutable getters
- multiple getters for the same field
- getter `with` closure attribute syntax
